XamarinCommunityToolkit/Views/GravatarImageSource/GravatarHandlerUtil.shared.cs

```diff
@@ -0,0 +1,100 @@
+#if !NETSTANDARD
+using System;
+using System.IO;
+using System.Net.Http;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using Xamarin.Forms;
+
+namespace Xamarin.CommunityToolkit.UI.Views
+{
+	static class GravatarHandlerUtil
+	{
+		const string requestUriFormat = "https://www.gravatar.com/avatar/{0}?s={1}&d={2}";
+		static readonly Lazy<HttpClient> lazyHttp = new Lazy<HttpClient>(() => new HttpClient());
+
+		public static async Task<FileInfo> Load(ImageSource imageSource, float scale, string cacheDirectory)
+		{
+			if (imageSource is GravatarImageSource gis)
+			{
+				var cacheFilePath = Path.Combine(cacheDirectory, nameof(GravatarImageSource), CacheFileName(gis, scale));
+				var cacheFileInfo = new FileInfo(cacheFilePath);
+
+				if (!UseCacheFile(gis.CachingEnabled, gis.CacheValidity, cacheFileInfo))
+				{
+					var imageBytes = await GetGravatarAsync(gis.Email, gis.Size, scale, gis.Default).ConfigureAwait(false);
+					SaveImage(cacheFileInfo, imageBytes);
+				}
+
+				return cacheFileInfo;
+			}
+
+			return null;
+		}
+
+		static void SaveImage(FileInfo cacheFileInfo, byte[] imageBytes)
+		{
+			if (imageBytes.Length < 1)
+				return;
+
+			// Delete Cached File
+			if (cacheFileInfo.Exists)
+				cacheFileInfo.Delete();
+
+			File.WriteAllBytes(cacheFileInfo.FullName, imageBytes);
+		}
```

Comment:
Should we add some "locks" to this file?

Reply:
other than the sample I would say it would be quite unusual to ever have more than one GravatarImageSource on a page with the same email really the worst that is going to happen is that you download it twice... for reading the file we shouldn't need a lock

Reply:
Yeah, I see.
But in case of slow internet and navigating from one page (Overview) to another one (Details) we can cause a situation when two threads are writing bytes to the same file.

Reply:
this has been refactored and there are now locks around all file IO operations